diff --git a/extras/sntp/sntp.c b/extras/sntp/sntp.c
index 26ac110..885c610 100644
--- a/extras/sntp/sntp.c
+++ b/extras/sntp/sntp.c
@@ -300,6 +300,8 @@ static ip_addr_t sntp_last_server_address;
 static u32_t sntp_last_timestamp_sent[2];
 #endif /* SNTP_CHECK_RESPONSE >= 2 */
 
+static time_t lastUpdated = 0;
+
 /**
  * SNTP processing of received timestamp
  */
@@ -324,6 +326,8 @@ sntp_process(u32_t *receive_timestamp)
   /* display local time from GMT time */
   LWIP_DEBUGF(SNTP_DEBUG_TRACE, ("sntp_process: %s", ctime(&t)));
 #endif /* SNTP_CALC_TIME_US */
+  
+  time(&lastUpdated);
 }
 
 /**
@@ -745,4 +749,10 @@ void sntp_set_update_delay(uint32_t ms)
 	sntp_update_delay = ms > 15000?ms:15000;
 }
 
+uint32_t sntp_get_last_update_age_sec(){
+	time_t ts;
+	time(&ts);
+	return ts - lastUpdated;
+}
+
 #endif /* LWIP_UDP */
diff --git a/extras/sntp/sntp.h b/extras/sntp/sntp.h
index 4fde486..731df79 100644
--- a/extras/sntp/sntp.h
+++ b/extras/sntp/sntp.h
@@ -72,5 +72,10 @@ time_t sntp_get_rtc_time(int32_t *us);
  */
 void sntp_update_rtc(time_t t, uint32_t us);
 
+/* 
+ *Get te number of seconds that have passed since the last update
+ */
+uint32_t sntp_get_last_update_age_sec();
+
 #endif /* _SNTP_H_ */
 
diff --git a/include/ssid_config.h b/include/ssid_config.h
index 9136731..b0488a9 100644
--- a/include/ssid_config.h
+++ b/include/ssid_config.h
@@ -13,6 +13,8 @@
 #ifndef __SSID_CONFIG_H__
 #define __SSID_CONFIG_H__
 
-#include "private_ssid_config.h"
+//#include "private_ssid_config.h"
+#define WIFI_SSID "NOT_USED"
+#define WIFI_PASS "NOT_USED"
 
 #endif // __SSID_CONFIG_H__
